inference_overview.md

```diff
@@ -0,0 +1,449 @@
+  as input to `generateBid()`. Ad embeddings can be included as a part of this
+  response. The Bidding service then invokes `generateBid()`. This call includes
+  the `buyer_signals` and data from the Ad retrieval server. Thus, contextual
+  embedding and version (part of `buyer_signals`) and the ad embeddings (part of
+  data looked up from the Ad retrieval server) are sent to the `generateBid()`
+  invocation.
+1. The calculated bid is returned. To compute the bid, `generateBid()` can run
+  models through `runInference()calls` and combine the output embedding with
+  other embeddings that were fetched above to make a final prediction.
+
+In the above flow, `prepareDataForAdRetrieval()` and `generateBid()` can use
+external models through the `runInference()` API call. This runs inside the TEE
+and can perform inference with models using private features (e.g. the user
+tower). Alternatively (not shown in figure), each of the UDFs can use embedded
+models to perform inference.
+
+## Versioning
+
+Inference on B&A will support factorized models. There are multiple models and
+embeddings that contribute towards the final prediction, some of which may be
+pre-computed. It is important to make sure that model and embedding versions
+used are consistent.
+
+For facilitating this, we will provide the buyer a way to choose the version of
+models to run per request. As a part of contextual request processing, a version
+for inference can be decided by the buyer. The model version information can be
+passed through the `buyer_signals`. This field is a string and can be used to
+pass arbitrary information from the contextual RTB request to `generateBid().
+buyer_signals` will also be passed to the K/V server where this model version
+information can be used to look up embeddings. The ad tech is free to choose the
+structure of the `buyer_signals` including how the model information is passed
+and what it contains. For example, something like the following could be used:
+
+### Using versions in B&A
+
+The `runInference()` API runs inferences against a named model. `generateBid()`
+is responsible for interpreting the version included in `buyer_signals` and
+translating it to an actual path that corresponds to an entity in the cloud
+bucket. A simple translation scheme could be something like
+`${model_name}-${model_version`}. This name will be translated into a cloud
+bucket folder from which the model was loaded as shown in the figure below. This
+figure is for illustrative purposes only. It shows the mapping from model path
+to an actual path in the cloud bucket. The cloud bucket will only be synced
+periodically, so each `runInference()` call will not call out to the cloud
+bucket. Thus, there will be some delay between a new model being available in
+the cloud bucket, and it being made available to `runInference()`. We will cover
+such details in future updates to this explainer.
+
+<figure id = "image-4">
```

Comment:
Can we remove this diagram in this explainer given that we said "details in future updates" ?

Reply:
I think the diagram is helpful to get a mental image of what is happening - this is the scope and purpose of this explainer. We can go into detailed explanations of how this happens in the future.